docs/spec_draft.md

```diff
@@ -604,6 +605,86 @@ Afterwards, within each `process_group`:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.all_reduce
+
+### Semantics
+
+Within each process group in the StableHLO grid, performs reduction, using
+`computations`, over the  values of the `operand` tensor from each process and
+produces a `result` tensor.
+
+The operation splits the StableHLO grid into process groups as follows:
+  * `channel_id <= 0` and `use_global_device_ids = false`,
+    `cross_replica(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = false`,
+    `cross_replica_and_partition(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = true`,
+    `flattened_ids(replica_groups)`.
+
+Afterwards, within each `process_group`:
+  * `operands@receiver = [operand@sender for sender in process_group]` for all
+    `receiver` in `process_group`.
+  * ```
+    result@process[i0, i1, ..., iR-1] =
+        reduce(
+          inputs=[operands@process[j][i0, i1, ..., iR-1] for j in range(size(operands@process))],
+          init_values=[I], # such that `computation` and `I` form a monoid]
+          dimensions=[0],
+          body=computation
+        )
+    ```
```

Comment:
Proposed reformulation

```
result@process[i0, i1, ..., iR-1] =
    reduce_without_init(
      inputs=operands@process[:][i0, i1, ..., iR-1],
      dimensions=[0],
      body=computation
    )

where `reduce_without_init` works exactly like `reduce`, except that its
`schedule` doesn't include init values.
```

Feedback:
  1. Used the slicing syntax to compress the for comprehension.
  2. Dropped "init_values=[I], # such that `computation` and `I` form a monoid]", because I don't think it's possible to determine I in the general case.